sdk/identity/identity-broker/test/manual/node/popTokenAuthenticationPolicy.ts

```diff
@@ -0,0 +1,223 @@
+   * The request may be modified before being sent.
+   */
+  authorizeRequestOnChallenge?(options: AuthorizeRequestOnChallengeOptions): Promise<any>;
+}
+
+/**
+ * Options to configure the popTokenAuthenticationPolicy
+ */
+export interface PopTokenAuthenticationPolicyOptions {
+  /**
+   * The TokenCredential implementation that can supply the pop token.
+   */
+  credential?: TokenCredential;
+  /**
+   * The scopes for which the pop token applies.
+   */
+  scopes: string | string[];
+  /**
+   * Allows for the processing of [Continuous Access Evaluation](https://docs.microsoft.com/azure/active-directory/conditional-access/concept-continuous-access-evaluation) challenges.
+   * If provided, it must contain at least the `authorizeRequestOnChallenge` method.
+   * If provided, after a request is sent, if it has a challenge, it can be processed to re-send the original request with the relevant challenge information.
+   */
+  challengeCallbacks?: ChallengeCallbacks;
+  /**
+   * A logger can be sent for debugging purposes.
+   */
+  logger?: AzureLogger;
+}
+
+/**
+ * Default authorize request handler
+ */
+async function defaultAuthorizeRequest(options: AuthorizeRequestOptions): Promise<void> {
+  const { scopes, getAccessToken, request } = options;
+  const getTokenOptions: GetTokenOptions = {
+    abortSignal: request.abortSignal,
+    tracingOptions: request.tracingOptions,
+  };
+
+  const accessToken = await getAccessToken(scopes, getTokenOptions);
+
+  if (accessToken) {
+    options.request.headers.set("Authorization", `pop ${accessToken.token}`);
+  }
+}
+
+/**
+ * We will retrieve the challenge only if the response status code was 401,
+ * and if the response contained the header "WWW-Authenticate" with a non-empty value.
+ */
+function getChallenge(response: PipelineResponse): string | undefined {
+  const challenge = response.headers.get("WWW-Authenticate");
+  if (response.status === 401 && challenge) {
+    return challenge;
+  }
+  return;
+}
+
+/**
+ * A policy that can request a token from a TokenCredential implementation and
+ * then apply it to the Authorization header of a request as a pop token.
+ */
+export function popTokenAuthenticationPolicy(
+  options: PopTokenAuthenticationPolicyOptions,
+): PipelinePolicy {
+  const { credential, scopes, challengeCallbacks } = options;
+  const logger = options.logger;
+  const callbacks = {
+    authorizeRequest: challengeCallbacks?.authorizeRequest ?? defaultAuthorizeRequest,
+    authorizeRequestOnChallenge: challengeCallbacks?.authorizeRequestOnChallenge,
+    // keep all other properties
+    ...challengeCallbacks,
+  };
+
+  // This function encapsulates the entire process of reliably retrieving the token
+  // The options are left out of the public API until there's demand to configure this.
+  // Remember to extend `popTokenAuthenticationPolicyOptions` with `TokenCyclerOptions`
+  // in order to pass through the `options` object.
+  const getAccessToken = credential
```

Comment:
I feel like we should _vastly_ simplify this manual test - all the tokenCycler logic isn't really needed - you can call `getAccessToken` every time since this is a test scenario. Testing all the tokenCycler logic is not really relevant.

I think we should make this _as simple as possible_ - removing any challenge stuff, token cycler stuff, anything that is not going to directly help with "I can make a single call to graph using pop"

Reply:
- We need the challenge logic for the nonce

Reply:
A pop token contains 2 calls - one to get the nonce and the other uses the nonce in the challenge to get the actual access token